tools/sync/command.py

```diff
@@ -61,6 +71,41 @@ def execute(self):
         Execute the command and capture its output and return code.
         """
 
+        class TimeoutThread(threading.Thread):
+            """
+            Wait until the timeout specified in seconds expires and kill
+            the process specified by the Popen object after that.
+            If timeout expires, TimeoutException is stored in the object
+            and can be retrieved by the caller.
+            """
+
+            def __init__(self, logger, timeout, condition, p):
+                super(TimeoutThread, self).__init__()
+                self.timeout = timeout
+                self.popen = p
+                self.condition = condition
+                self.logger = logger
+                self.start()
+                self.exception = None
+
+            def run(self):
+                with self.condition:
+                    if not self.condition.wait(self.timeout):
+                        p = self.popen
+                        self.logger.info("Terminating command {} with PID {} "
+                                         "after timeout of {} seconds".
+                                         format(p.args, p.pid, self.timeout))
+                        self.popen.terminate()
+                        self.exception = TimeoutException("Command {} with pid"
+                                                          " {} timed out".
+                                                          format(p.args,
+                                                                 p.pid))
+                    else:
+                        return None
+
+            def getexception(self):
```

Comment:
Maybe rename to `get_exception` or `getException`? (Not sure what Python style is.)

Reply:
`get_exception`

Reply:
ok